src/py/workflows/nextstrain_run/save.py

```diff
@@ -0,0 +1,149 @@
+@click.option("--tree-path", type=click.File("r"), required=True)
+def cli(
+    aspen_workflow_rev: str,
+    aspen_creation_rev: str,
+    ncov_rev: str,
+    aspen_docker_image_version: str,
+    end_time: int,
+    phylo_run_id: int,
+    bucket: str,
+    key: str,
+    tree_path: io.TextIOBase,
+):
+    end_time_datetime = datetime.datetime.fromtimestamp(end_time)
+
+    interface: SqlAlchemyInterface = init_db(get_db_uri(RemoteDatabaseConfig()))
+    with session_scope(interface) as session:
+        phylo_run: PhyloRun = (
+            session.query(PhyloRun)
+            .filter(PhyloRun.workflow_id == phylo_run_id)
+            .options(
+                joinedload(PhyloRun.inputs.of_type(UploadedPathogenGenome))
+                # load the sample that this uploaded pathogen genome was associated with
+                .subqueryload(UploadedPathogenGenome.sample),
+                joinedload(PhyloRun.inputs.of_type(CalledPathogenGenome))
+                # load the workflows that generated the consensus calls.
+                .subqueryload(
+                    CalledPathogenGenome.producing_workflow.of_type(CallConsensus)
+                )
+                # load the host-filtered sequencing reads that generated the called
+                # pathogen genomes.
+                .subqueryload(
+                    CallConsensus.inputs.of_type(HostFilteredSequencingReadsCollection)
+                )
+                # load the workflows that generated the host-filtered sequencing
+                # reads.
+                .subqueryload(
+                    HostFilteredSequencingReadsCollection.producing_workflow.of_type(
+                        FilterRead
+                    )
+                )
+                # load the sequencing reads that generated the host-filtered
+                # sequencing reads.
+                .subqueryload(FilterRead.inputs.of_type(SequencingReadsCollection))
+                # load the samples that the sequencing reads collection was associated
+                # with.
+                .subqueryload(SequencingReadsCollection.sample),
+            )
+            .one()
+        )
+
+        assert phylo_run.workflow_status != WorkflowStatusType.COMPLETED
+
+        # read the tree
+        tree_json = json.load(tree_path)
+        all_public_identifiers = get_names_from_tree([tree_json["tree"]])
+
+        # get all the children that are pathogen genomes
+        pathogen_genomes = [
+            inp for inp in phylo_run.inputs if isinstance(inp, PathogenGenome)
+        ]
+
+        uploaded_pathogen_genomes: Set[UploadedPathogenGenome] = {
+            pathogen_genome
+            for pathogen_genome in pathogen_genomes
+            if isinstance(pathogen_genome, UploadedPathogenGenome)
+        }
+        sequencing_reads_collections: Set[SequencingReadsCollection] = {
+            pathogen_genome.get_parents(HostFilteredSequencingReadsCollection)[
+                0
```

Comment:
lol at this formatting

Reply:
Yeah, I very much disagree with black on this.